### PR TITLE
Label for Pull Request and Issue

### PR DESCRIPTION
**Pull request by crimsonfaith91**
Implementing the label feature for pull request and issue.

**Testing done:**
(1) all unit test cases written by Garrett were passed
(2) webhook API call:
     - issue: 200 response with no message shown on cmd/hook console (test_issue.json)
     - pull request: 200 response with a message shown on cmd/hook console (pr.json)
*Note: I created two new JSON files with my own credentials and labels (e.g. /area, /kind) based on this issue: https://github.com/kubernetes/test-infra/issues/2445

This is my first PR. I apologize if I conducted a lot of mistakes or left out some corner cases. Feel free to criticize my codes. Thanks for your patience!

### CODE REVIEW

**Comment by spxtr** [comment]
Ah, I see what you mean. Check out the [assign plugin](https://github.com/kubernetes/test-infra/blob/master/prow/plugins/assign/assign.go) for an example of how to refactor this.

**Comment by crimsonfaith91** [comment]
Noted! Let me have a look at it. Thanks!

**Comment by crimsonfaith91** [comment]
Updated cla/linuxfoundation requirement. The second commit now shows me as the author.

**Comment by spxtr** [comment]
Please redo the first commit so that it passes CLA as well, or else we can't merge!

File: prow/plugins/label/label.go [hunk 1]
```diff
@@ -37,6 +49,8 @@ var (
 
 func init() {
 	plugins.RegisterIssueCommentHandler(pluginName, handleIssueComment)
+	plugins.RegisterIssueHandler(pluginName, handleIssue)
+	//plugins.RegisterPullRequestHandler(pluginName, handlePullRequest)
```
Comment (spxtr) [inline prow/plugins/label/label.go]: Probably this shouldn't be commented out.

File: prow/plugins/label/label.go [hunk 2]
```diff
@@ -109,13 +161,14 @@ func handle(gc githubClient, log *logrus.Entry, ic github.IssueCommentEvent) err
 		}
 
 	}
-
+	
 	if len(nonexistent) > 0 {
 		msg := fmt.Sprintf(nonExistentLabel, strings.Join(nonexistent, ", "))
-		if err := gc.CreateComment(owner, repo, number, plugins.FormatICResponse(ic.Comment, msg)); err != nil {
+		if err := gc.CreateComment(owner, repo, number, plugins.FormatICResponse(ae.comment, msg)); err != nil {
+		//if err := gc.CreateComment(owner, repo, number, plugins.FormatResponseRaw(ae.body, ae.url, ae.login, msg)); err != nil {
```
Comment (spxtr) [inline prow/plugins/label/label.go]: ?

File: prow/plugins/label/label_test.go [hunk 3]
```diff
@@ -20,10 +20,11 @@ import (
 	"fmt"
 	"testing"
 
+	"reflect"
```
Comment (spxtr) [inline prow/plugins/label/label_test.go]: In general it's bad form to use the reflect package at all, including in tests. Sometimes I allow it just because it's easier, but it's a bad sign.

File: prow/plugins/label/label_test.go [hunk 4]
```diff
@@ -34,6 +35,15 @@ const (
 	prNumber     = 1
 )
 
+type TestCase struct {
```
Comment (spxtr) [inline prow/plugins/label/label_test.go]: `testCase`

File: prow/plugins/label/label_test.go [hunk 5]
```diff
@@ -241,19 +364,13 @@ func TestLabel(t *testing.T) {
 	}
 
 	for _, tc := range testcases {
-		fakeClient, ice := getFakeRepo(tc.body, tc.commenter, tc.repoLabels)
-		if err := handle(fakeClient, logrus.WithField("plugin", pluginName), ice); err != nil {
-			t.Errorf("For case %s, didn't expect error from label test: %v", tc.name, err)
-			continue
-		}
+		fakeClient, ice := getFakeRepoIssueComment(tc.body, tc.commenter, tc.repoLabels) //issue comment
```
Comment (spxtr) [inline prow/plugins/label/label_test.go]: You don't need the comment at the end.

Comment (spxtr) [inline prow/plugins/label/label_test.go]: Here and below.

**Comment by spxtr** [comment]
@k8s-bot ok to test

File: prow/plugins/label/label.go (hunk 1 above)
Comment (crimsonfaith91) [inline prow/plugins/label/label.go]: Yes!

File: prow/plugins/label/label.go (hunk 2 above)
Comment (crimsonfaith91) [inline prow/plugins/label/label.go]: "FormatResponseRaw" may be better than "FormatICResponse" as we now cater for issue comment, issue, and pull request.

File: prow/plugins/label/label_test.go (hunk 3 above)
Comment (crimsonfaith91) [inline prow/plugins/label/label_test.go]: i will remove it.

File: prow/plugins/label/label_test.go (hunk 4 above)
Comment (crimsonfaith91) [inline prow/plugins/label/label_test.go]: ok

File: prow/plugins/label/label_test.go (hunk 5 above)
Comment (crimsonfaith91) [inline prow/plugins/label/label_test.go]: noted

**Comment by crimsonfaith91** [comment]
Sorry, i should not close this PR.
I tried to get rid of irrelevant commits using "git reset --hard".
I was not aware it would close the PR if the PR has no commit left...
I will submit another PR or reopen this PR.

**Comment by crimsonfaith91** [comment]
The Travis Cl build failed due to a connection timeout when finding an image. Should we re-run the build?

**Comment by crimsonfaith91** [comment]
@foxish 
@grodrigues3

**Comment by grodrigues3** [comment]
@k8s-bot ok to test

**Comment by grodrigues3** [comment]
@k8s-bot test this

**Comment by spxtr** [comment]
Ok, it looks like this is going to need a nontrivial rebase after #2456, which I think we're going to do first.

File: prow/plugins/label/label_test.go [hunk 6]
```diff
@@ -70,20 +79,141 @@ func getFakeRepo(commentBody, commenter string, repoLabels []string) (*fakegithu
...
+		comment: github.IssueComment{
+			Body: pre.PullRequest.Body,
+			User: pre.PullRequest.User,
+		},
+	}
+
+	return fakeCli, ae
+}
+
+func examineTestCase(t *testing.T, fakeClient *fakegithub.FakeClient, ae assignEvent, tc testCase) {
```
Comment (foxish) [inline prow/plugins/label/label_test.go]: We should have a more descriptive name for this method here I think. `testLabels` or something maybe?

Comment (crimsonfaith91) [inline prow/plugins/label/label_test.go]: sure! i will change it now. :)

**Comment by spxtr** [comment]
OK, merged #2456. Please rebase this and I'll take another look.

File: prow/plugins/label/label_test.go (hunk 6 above)
Comment (foxish) [inline prow/plugins/label/label_test.go]: The plural "labels" refers to the fact that we're testing expectations with regard to labels in this method. I haven't seen the tests in detail but If we're adding and testing for exactly one label addition, then testLabel works as well.

Comment (crimsonfaith91) [inline prow/plugins/label/label_test.go]: I got what you mean. Some test cases involve more than one label, so it is better to make it plural. Thanks! :)

**Comment by crimsonfaith91** [comment]
I deleted my remote branch to clean up my workspace, but I didn't expect this PR will be closed automatically... From now on, please refer to: https://github.com/kubernetes/test-infra/pull/2505

Thanks!
